### ISSUE TITLE
[Discussion]  How to inference with batch sample

### ISSUE THREAD

**Issue by yuntaodu** [haotian-liu/LLaVA#407]
### Discussion

Hi, now we could input one single image and question to the model for evaluation. What if we want to test the mode with a batch of samples, What should we do?

**Comment by LouieBHLu** [comment]
I have the same question.

**Comment by basteran** [comment]
Same here, I cannot find anything around.. I guess we need to develop our own code, but I can't figure out how to initialize the model alone with a script.. any suggestions?

**Comment by rabiulcste** [comment]
The snippet below should work on batched inference. 


First, define define a processor, get batched `input_ids` and `image_tensor`. Now, you can do inference!

```
class LlaVaProcessor:
    def __init__(self, tokenizer, image_processor, mm_use_im_start_end):
        self.mm_use_im_start_end = mm_use_im_start_end
        self.tokenizer = tokenizer
        self.image_processor = image_processor
        self.conv_mode = "llava_v1"

    def format_text(self, text: str):
        if self.mm_use_im_start_end:
            text = DEFAULT_IM_START_TOKEN + DEFAULT_IMAGE_TOKEN + DEFAULT_IM_END_TOKEN + "\n" + text
        else:
            text = DEFAULT_IMAGE_TOKEN + "\n" + text

        conv = conv_templates[self.conv_mode].copy()
        conv.append_message(conv.roles[0], text)
        conv.append_message(conv.roles[1], None)
        text = conv.get_prompt()

        return text

    def load_image(self, image_path: str):
        return Image.open(image_path).convert("RGB")

    @staticmethod
    def pad_sequence_to_max_length(sequence, max_length, padding_value=0):
        """Pad a sequence to the desired max length."""
        if len(sequence) >= max_length:
            return sequence
        return torch.cat([sequence, torch.full((max_length - len(sequence),), padding_value, dtype=sequence.dtype)])

    def get_processed_tokens(self, text: str, image_path: str):
        prompt = self.format_text(text)
        image = self.load_image(image_path)

        input_ids = tokenizer_image_token(prompt, self.tokenizer, IMAGE_TOKEN_INDEX, return_tensors="pt").unsqueeze(0)
        image_tensor = self.image_processor.preprocess(image, return_tensors="pt")["pixel_values"][0]

        return image_tensor, input_ids

    def get_processed_tokens_batch(self, batch_text: List[str], image_paths: List[str]):
        prompt = [self.format_text(text) for text in batch_text]
        images = [self.load_image(image_path) for image_path in image_paths]

        batch_input_ids = [
            tokenizer_image_token(prompt, self.tokenizer, IMAGE_TOKEN_INDEX, return_tensors="pt") for prompt in prompt
        ]

        # Determine the maximum length of input_ids in the batch
        max_len = max([len(seq) for seq in batch_input_ids])
        # Pad each sequence in input_ids to the max_len
        padded_input_ids = [self.pad_sequence_to_max_length(seq.squeeze(), max_len) for seq in batch_input_ids]
        batch_input_ids = torch.stack(padded_input_ids)

        batch_image_tensor = self.image_processor(images, return_tensors="pt")["pixel_values"]

        return batch_image_tensor, batch_input_ids

```
```
# get tokenized data as batch
image_tensor, input_ids = processor.get_processed_tokens_batch(
    [example["prompted_question"] for example in batch], [example["image_path"] for example in batch]
)
```

```
# run inference
conv = conv_templates[processor.conv_mode].copy()
stop_str = conv.sep if conv.sep_style != SeparatorStyle.TWO else conv.sep2
keywords = [stop_str]
stopping_criteria = (
    [KeywordsStoppingCriteria(keywords, tokenizer_, input_ids)] if conv.version == "v0" else None
)
input_ids = batch["input_ids"]
image_tensor = batch["image_tensors"]
input_ids = input_ids.cuda()

with torch.inference_mode():
    output_ids = model.generate(
        input_ids,
        images=image_tensor.half().cuda(),
        do_sample=False,
        num_beams=args.num_beams,
        max_new_tokens=args.max_length,
        length_penalty=args.length_penalty,
        use_cache=True,
        stopping_criteria=stopping_criteria,
    )
generated_outputs = tokenizer_.batch_decode(output_ids[:, input_ids.shape[1]:], skip_special_tokens=True)
generated_outputs = [out.strip() for out in generated_outputs]
generated_outputs = [out[:-len(stop_str)] if out.endswith(stop_str) else out for out in generated_outputs]
```

Hope it helps!

**Comment by rabiulcste** [comment]
@basteran It should work as the case with any transformer model. You need to do some batch processing to feed the model. However, I don't see any inference speed with batch inference -- which is very odd! @haotian-liu any thoughts on the speed? 
 
```
# define a processor

class LlaVaProcessor:
  def __init__(self, tokenizer, image_processor, mm_use_im_start_end):
      self.mm_use_im_start_end = mm_use_im_start_end
      self.tokenizer = tokenizer
      self.image_processor = image_processor
      self.conv_mode = "llava_v1"

  def format_text(self, text: str):
      if self.mm_use_im_start_end:
          text = DEFAULT_IM_START_TOKEN + DEFAULT_IMAGE_TOKEN + DEFAULT_IM_END_TOKEN + "\n" + text
      else:
          text = DEFAULT_IMAGE_TOKEN + "\n" + text

      conv = conv_templates[self.conv_mode].copy()
      conv.append_message(conv.roles[0], text)
      conv.append_message(conv.roles[1], None)
      text = conv.get_prompt()

      return text

  def load_image(self, image_path: str):
      return Image.open(image_path).convert("RGB")

  @staticmethod
  def pad_sequence_to_max_length(sequence, max_length, padding_value=0):
      """Pad a sequence to the desired max length."""
      if len(sequence) >= max_length:
          return sequence
      return torch.cat([sequence, torch.full((max_length - len(sequence),), padding_value, dtype=sequence.dtype)])

  def get_processed_tokens(self, text: str, image_path: str):
      prompt = self.format_text(text)
      image = self.load_image(image_path)

      input_ids = tokenizer_image_token(prompt, self.tokenizer, IMAGE_TOKEN_INDEX, return_tensors="pt").unsqueeze(0)
      image_tensor = self.image_processor.preprocess(image, return_tensors="pt")["pixel_values"][0]

      return image_tensor, input_ids

  def get_processed_tokens_batch(self, batch_text: List[str], image_paths: List[str]):
      prompt = [self.format_text(text) for text in batch_text]
      images = [self.load_image(image_path) for image_path in image_paths]

      batch_input_ids = [
          tokenizer_image_token(prompt, self.tokenizer, IMAGE_TOKEN_INDEX, return_tensors="pt") for prompt in prompt
      ]

      # Determine the maximum length of input_ids in the batch
      max_len = max([len(seq) for seq in batch_input_ids])
      # Pad each sequence in input_ids to the max_len
      padded_input_ids = [self.pad_sequence_to_max_length(seq.squeeze(), max_len) for seq in batch_input_ids]
      batch_input_ids = torch.stack(padded_input_ids)

      batch_image_tensor = self.image_processor(images, return_tensors="pt")["pixel_values"]

      return batch_image_tensor, batch_input_ids
```

```
# load models and tokenizers

from dataset_zoo.custom_processor import LlaVaProcessor
from LLaVA.llava.conversation import SeparatorStyle, conv_templates
from LLaVA.llava.mm_utils import (KeywordsStoppingCriteria,
                                  get_model_name_from_path)
from LLaVA.llava.model.builder import load_pretrained_model
from LLaVA.llava.utils import disable_torch_init

disable_torch_init()
model_path = os.path.expanduser(model_name)
model_name = get_model_name_from_path(model_path)
model_name = get_model_name_from_path(model_path)
model_base = MODEL_CLS_INFO["hfformer"][args.model_name]["base"]
tokenizer_, model, image_processor, context_len = load_pretrained_model(model_path, model_base, model_name)
tokenizer_.padding_side = "left"
processor = LlaVaProcessor(tokenizer_, image_processor, model.config.mm_use_im_start_end)
```

```
batch_images, batch_text = processor.get_processed_tokens_batch(
    [example["prompted_question"] for example in batch], [example["image_path"] for example in batch]
)
processed_batch["image_tensors"] = batch_images
processed_batch["input_ids"] = batch_text

# run batch inference
conv = conv_templates[processor.conv_mode].copy()
stop_str = conv.sep if conv.sep_style != SeparatorStyle.TWO else conv.sep2
keywords = [stop_str]
stopping_criteria = (
    [KeywordsStoppingCriteria(keywords, tokenizer_, input_ids)] if conv.version == "v0" else None
)
input_ids = batch["input_ids"]
image_tensor = batch["image_tensors"]
input_ids = input_ids.cuda()

with torch.inference_mode():
    output_ids = model.generate(
        input_ids,
        images=image_tensor.half().cuda(),
        do_sample=False,
        num_beams=args.num_beams,
        max_new_tokens=args.max_length,
        length_penalty=args.length_penalty,
        use_cache=True,
        stopping_criteria=stopping_criteria,
    )
generated_outputs = tokenizer_.batch_decode(output_ids[:, input_ids.shape[1] :], skip_special_tokens=True)
generated_outputs = [out.strip() for out in generated_outputs]
generated_outputs = [out[: -len(stop_str)] if out.endswith(stop_str) else out for out in generated_outputs]
```

Hope it helps!

**Comment by NinaadRao** [comment]
When I try to use batch for inferencing, I get an issue where stopping_criteria expects batch size as 1. Is there any specific reason to this?

**Comment by dverdu-freepik** [comment]
Hi @rabiulcste, in your code there is a mistake, as you should use left padding

```python
@staticmethod
def pad_sequence_to_max_length(sequence, max_length, padding_value=0):
      """Pad a sequence to the desired max length."""
      if len(sequence) >= max_length:
          return sequence
      return torch.cat([torch.full((max_length - len(sequence),), padding_value, dtype=sequence.dtype), sequence])
```

**Comment by haotian-liu** [comment]
Hi, thank you, and I am trying to work on an efficient batch inference. Closing this and consolidating the discussion to https://github.com/haotian-liu/LLaVA/issues/754.

@dverdu-freepik @rabiulcste If you have prior experience in the batch inference, it is very much appreciated if you could help share some insights on why the batch inference I implemented were not efficient. Thanks.

**Comment by BeachWang** [comment]
> @basteran It should work as the case with any transformer model. You need to do some batch processing to feed the model. However, I don't see any inference speed with batch inference -- which is very odd! @haotian-liu any thoughts on the speed?
> 
> ```
> # define a processor
> 
> class LlaVaProcessor:
>   def __init__(self, tokenizer, image_processor, mm_use_im_start_end):
>       self.mm_use_im_start_end = mm_use_im_start_end
>       self.tokenizer = tokenizer
>       self.image_processor = image_processor
>       self.conv_mode = "llava_v1"
> 
>   def format_text(self, text: str):
>       if self.mm_use_im_start_end:
>           text = DEFAULT_IM_START_TOKEN + DEFAULT_IMAGE_TOKEN + DEFAULT_IM_END_TOKEN + "\n" + text
>       else:
>           text = DEFAULT_IMAGE_TOKEN + "\n" + text
> 
>       conv = conv_templates[self.conv_mode].copy()
>       conv.append_message(conv.roles[0], text)
>       conv.append_message(conv.roles[1], None)
>       text = conv.get_prompt()
> 
>       return text
> 
>   def load_image(self, image_path: str):
>       return Image.open(image_path).convert("RGB")
> 
>   @staticmethod
>   def pad_sequence_to_max_length(sequence, max_length, padding_value=0):
>       """Pad a sequence to the desired max length."""
>       if len(sequence) >= max_length:
>           return sequence
>       return torch.cat([sequence, torch.full((max_length - len(sequence),), padding_value, dtype=sequence.dtype)])
> 
>   def get_processed_tokens(self, text: str, image_path: str):
>       prompt = self.format_text(text)
>       image = self.load_image(image_path)
> 
>       input_ids = tokenizer_image_token(prompt, self.tokenizer, IMAGE_TOKEN_INDEX, return_tensors="pt").unsqueeze(0)
>       image_tensor = self.image_processor.preprocess(image, return_tensors="pt")["pixel_values"][0]
> 
>       return image_tensor, input_ids
> 
>   def get_processed_tokens_batch(self, batch_text: List[str], image_paths: List[str]):
>       prompt = [self.format_text(text) for text in batch_text]
>       images = [self.load_image(image_path) for image_path in image_paths]
> 
>       batch_input_ids = [
>           tokenizer_image_token(prompt, self.tokenizer, IMAGE_TOKEN_INDEX, return_tensors="pt") for prompt in prompt
>       ]
> 
>       # Determine the maximum length of input_ids in the batch
>       max_len = max([len(seq) for seq in batch_input_ids])
>       # Pad each sequence in input_ids to the max_len
>       padded_input_ids = [self.pad_sequence_to_max_length(seq.squeeze(), max_len) for seq in batch_input_ids]
>       batch_input_ids = torch.stack(padded_input_ids)
> 
>       batch_image_tensor = self.image_processor(images, return_tensors="pt")["pixel_values"]
> 
>       return batch_image_tensor, batch_input_ids
> ```
> 
> ```
> # load models and tokenizers
> 
> from dataset_zoo.custom_processor import LlaVaProcessor
> from LLaVA.llava.conversation import SeparatorStyle, conv_templates
> from LLaVA.llava.mm_utils import (KeywordsStoppingCriteria,
>                                   get_model_name_from_path)
> from LLaVA.llava.model.builder import load_pretrained_model
> from LLaVA.llava.utils import disable_torch_init
> 
> disable_torch_init()
> model_path = os.path.expanduser(model_name)
> model_name = get_model_name_from_path(model_path)
> model_name = get_model_name_from_path(model_path)
> model_base = MODEL_CLS_INFO["hfformer"][args.model_name]["base"]
> tokenizer_, model, image_processor, context_len = load_pretrained_model(model_path, model_base, model_name)
> tokenizer_.padding_side = "left"
> processor = LlaVaProcessor(tokenizer_, image_processor, model.config.mm_use_im_start_end)
> ```
> 
> ```
> batch_images, batch_text = processor.get_processed_tokens_batch(
>     [example["prompted_question"] for example in batch], [example["image_path"] for example in batch]
> )
> processed_batch["image_tensors"] = batch_images
> processed_batch["input_ids"] = batch_text
> 
> # run batch inference
> conv = conv_templates[processor.conv_mode].copy()
> stop_str = conv.sep if conv.sep_style != SeparatorStyle.TWO else conv.sep2
> keywords = [stop_str]
> stopping_criteria = (
>     [KeywordsStoppingCriteria(keywords, tokenizer_, input_ids)] if conv.version == "v0" else None
> )
> input_ids = batch["input_ids"]
> image_tensor = batch["image_tensors"]
> input_ids = input_ids.cuda()
> 
> with torch.inference_mode():
>     output_ids = model.generate(
>         input_ids,
>         images=image_tensor.half().cuda(),
>         do_sample=False,
>         num_beams=args.num_beams,
>         max_new_tokens=args.max_length,
>         length_penalty=args.length_penalty,
>         use_cache=True,
>         stopping_criteria=stopping_criteria,
>     )
> generated_outputs = tokenizer_.batch_decode(output_ids[:, input_ids.shape[1] :], skip_special_tokens=True)
> generated_outputs = [out.strip() for out in generated_outputs]
> generated_outputs = [out[: -len(stop_str)] if out.endswith(stop_str) else out for out in generated_outputs]
> ```
> 
> Hope it helps!

The same question.
